### PR TITLE
Code cleanup in the sync module

### PR DESCRIPTION
**Pull request by dvdplm**
Reading code, fixing small things as I went: a few clones less, mark test-only code as such, some docs, unused params.

### CODE REVIEW

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -2728,7 +2730,9 @@ impl ImportExportBlocks for Client {
 				}
 			}
 		};
-		self.flush_queue();
+		self.importer.block_queue.flush();
+		self.import_verified_blocks();
```
Comment (ordian) [inline ethcore/src/client/client.rs]: what is rationale for this change? the logic has changed here

File: ethcore/sync/src/blocks.rs [hunk 2]
```diff
@@ -317,42 +317,41 @@ impl BlockCollection {
 			return Vec::new();
 		}
 
-		let mut drained = Vec::new();
+		// todo[dvdplm]: should be possible to know how many these are.
```
Comment (ordian) [inline ethcore/sync/src/blocks.rs]: if there is no way, let's remove the todo

File: ethcore/sync/src/blocks.rs [hunk 3]
```diff
@@ -317,42 +317,41 @@ impl BlockCollection {
...
+		for block in blocks {
+			let unverified = unverified_from_sync(block.header, block.body);
+			drained.push(BlockAndReceipts {
+				block: unverified,
+				receipts: block.receipts,
+			});
 		}
 
-		trace!(target: "sync", "Drained {} blocks, new head :{:?}", drained.len(), self.head);
+		log::debug!(target: "sync", "Drained {} blocks, new head :{:?}", drained.len(), self.head);
```
Comment (ordian) [inline ethcore/sync/src/blocks.rs]: nit: let's stick to one style of import: either `debug!` or `log::debug!` everywhere

File: ethcore/src/client/client.rs (hunk 1 above)
Comment (dvdplm) [reply ethcore/src/client/client.rs]: > the logic has changed here

Not really? `flush_queue()` is not used outside tests except here. We know that we have just added blocks to the queue so no need to check it again.

File: ethcore/sync/src/blocks.rs (hunk 2 above)
Comment (dvdplm) [reply ethcore/sync/src/blocks.rs]: It's actually not trivial to find out the size of these collections, likely we keep blocks from all kinds of forks in the `self.blocks`/`self.parents` collections and there's no good way to anticipate how many blocks we're going to drain. Most of the time we end up draining 0 blocks and then we drain a whole bunch of them, up to 30 000. Not sure why this is tbh.

File: ethcore/sync/src/blocks.rs [hunk 4]
```diff
@@ -196,11 +196,11 @@ impl BlockCollection {
 	}
 
 	/// Insert a collection of block receipts for previously downloaded headers.
-	pub fn insert_receipts(&mut self, receipts: Vec<Bytes>) -> Vec<Vec<H256>> {
+	pub fn insert_receipts(&mut self, receipts: &[&[u8]]) -> Vec<Vec<H256>> {
```
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: :+1:

File: ethcore/sync/src/blocks.rs [hunk 5]
```diff
@@ -210,11 +210,11 @@ impl BlockCollection {
 	}
 
 	/// Returns a set of block hashes that require a body download. The returned set is marked as being downloaded.
-	pub fn needed_bodies(&mut self, count: usize, _ignore_downloading: bool) -> Vec<H256> {
```
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: This is great, boolean flags is such a pain IMO

File: ethcore/sync/src/blocks.rs [hunk 6]
```diff
@@ -210,11 +210,11 @@ impl BlockCollection {
...
+	pub fn needed_bodies(&mut self, count: usize) -> Vec<H256> {
 		if self.head.is_none() {
 			return Vec::new();
 		}
-		let mut needed_bodies: Vec<H256> = Vec::new();
+		let mut needed_bodies: Vec<H256> = Vec::with_capacity(count);
 		let mut head = self.head;
 		while head.is_some() && needed_bodies.len() < count {
 			head = self.parents.get(&head.unwrap()).cloned();
```
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: replace this with expect? 

`head is some checked above; qed`?

File: ethcore/sync/src/blocks.rs [hunk 7]
```diff
@@ -317,42 +317,40 @@ impl BlockCollection {
...
-							} else {
-								self.blocks.insert(head, block);
-								break;
-							}
-						},
-						_ => {
+		let mut blocks = Vec::new();
+		let mut head = self.head;
+		while let Some(h) = head {
+			head = self.parents.get(&h).cloned();
```
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: this is `H256` could be deref instead of clone.

File: ethcore/sync/src/blocks.rs (hunk 6 above)
Comment (ordian) [reply ethcore/sync/src/blocks.rs]: or with  `while let Some(head) = head`

File: ethcore/sync/src/blocks.rs [hunk 8]
```diff
@@ -422,7 +420,7 @@ impl BlockCollection {
 					match self.blocks.get_mut(&h) {
 						Some(ref mut block) => {
 							trace!(target: "sync", "Got receipt {}", h);
-							block.receipts = Some(r.clone());
+							block.receipts = Some(r.to_vec());
```
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: why not just move it and keep r: Bytes?

File: ethcore/sync/src/blocks.rs (hunk 7 above)
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: H256 is `Copy`, 

```suggestion
			head = self.parents.get(&h).copied();
```

File: ethcore/sync/src/blocks.rs (hunk 8 above)
Comment (ordian) [reply ethcore/sync/src/blocks.rs]: it is set in a loop, so we need to copy the data in any case

File: ethcore/sync/src/blocks.rs (hunk 6 above)
Comment (niklasad1) [reply ethcore/sync/src/blocks.rs]: I don't think that works because `head` is mutated in the loop?!

Comment (ordian) [reply ethcore/sync/src/blocks.rs]: alternatively
```rust
while needed_bodies.len() < count {
	let head = match head {
		Some(head) => self.parents.get(head).copied(),
		None => break,
	};
```

File: ethcore/sync/src/blocks.rs (hunk 4 above)
Comment (niklasad1) [inline ethcore/sync/src/blocks.rs]: hmm, I didn't understand this first.

Personally, I would prefer to have `Vec<Bytes>` here and just move each receipt in `fn insert_receipt` and don't clone it there if that's possible (seems so) but I don't have any strong opinions on it so do what you want.

File: ethcore/src/client/client.rs (hunk 1 above)
Comment (ordian) [reply ethcore/src/client/client.rs]: As discussed previosly `flush_queue` was doing `import_verified_blocks()` in a loop and we removed the logic here, so I think the method removal should be reverted.

Comment (ordian) [reply ethcore/src/client/client.rs]: or maybe I'm misunderstanding why it was done in a loop previously

Comment (niklasad1) [reply ethcore/src/client/client.rs]: ok I see, I thought all blocks were imported in `import_verified_blocks()` but it actually has an internal configuration, `max_blocks_to_import` so I think that this should be reverted too. Thus, so we are sure that the `block queue` actually is empty.

Comment (dvdplm) [reply ethcore/src/client/client.rs]: Yeah, I messed up here. 
I'm still annoyed by having a `pub` that is used only here and from a bunch of tests. I mistakenly thought it was an "important" call for the main importing code path but it's not. I'd probably prefer keeping it test-only and copy the logic here if I was to decide but I know ya'all would hate that so… reverting.

File: ethcore/sync/src/blocks.rs [hunk 9]
```diff
@@ -241,25 +245,28 @@ impl BlockCollection {
...
+								self.downloading_receipts.insert(block.receipts_root);
+								needed_receipts.push(head);
+							}
 						}
+						_ => (),
 					}
-					_ => (),
-				}
-			}
+					self.parents.get(&head.unwrap()).copied()
```
Comment (ordian) [inline ethcore/sync/src/blocks.rs]:
```suggestion
					self.parents.get(&head).copied()
```

**Pull request merged**